### PR TITLE
chore: add DotcomWeb.CoreComponents

### PR DESCRIPTION
**Pull request by bklebe**
#### Summary of changes

**Asana Ticket:** [Import the set of base `Phoenix.Components`](https://app.asana.com/0/555089885850811/1207828528645183/f)

Supersedes #2136. 

---

#### General checks
* [ ] **Are the changes organized into self-contained commits with descriptive and well-formatted commit messages?** This is a good practice that can facilitate easier reviews.
* [ ] **Testing.** Do the changes include relevant passing updates to tests? This includes updating screenshots. Preferably tests are run locally to verify that there are no test failures created by these changes, before opening a PR.
* [ ] **Tech debt.** Have you checked for tech debt you can address in the area you're working in? This can be a good time to address small issues, or create Asana tickets for larger issues.

#### New UI, or substantial UI changes
* [ ] **Cross-browser compatibility** is less of an issue now that we're no longer supporting IE, but changes still need to work as expected in Safari, Chrome, and Firefox.
* [ ] **Are interactive elements accessible?** This includes at minimum having [relevant keyboard interactions](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#keyboard-interaction) and [visible focus](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#visible-focus), but can also include verification with screen reader testing.
* [ ] **Other accessibility checks** such as sufficient [color constrast](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#color-contrast), or whether the layout holds up at 200% [zoom level](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#visible-focus).

#### New endpoints, or non-trivial changes to current endpoints
* [ ] **Have we load-tested** any new pages or internal API endpoints that will receive significant traffic? [See load testing docs](./../load_tests/README.md)
* [ ] **If this change involves routes,** does it work correctly with pertinent "unusual" routes such as the combined Green Line, Silver Line, Foxboro commuter rail, and single-direction bus routes like the 170?

### CODE REVIEW

**Comment by bklebe** [comment]
Not totally sure how to test this, there are at least no more Webpack errors. Let me know what you think —

**Review by thecristen — CHANGES_REQUESTED** [review]
super cool!
- can you please add `import DotcomWeb.CoreComponents` to the `DotcomWeb` `view_helpers` block? then the components will be easy to invoke anywhere!
- I tried out the `<.input />` and the styling doesn't seem to be quite working, I just get a mostly plain HTML output. Maybe we're needing an import statement somewhere to bring the Tailwind classes in? Or did you want to leave that for a followup PR?

**Comment by bklebe** [comment]
@thecristen 

> can you please add import DotcomWeb.CoreComponents to the DotcomWeb view_helpers block? then the components will be easy to invoke anywhere!

This sounds like a good idea, but it would require either renaming `icon` component in CoreComponents to `hero_icon` or something, or renaming all the other functions called `icon` that we use in views, e.g.: https://github.com/search?q=repo%3Ambta%2Fdotcom+def+icon&type=code. Let me know which one you want me to do, or whether I should leave it for a followup.

Still trying to figure out why Tailwind isn't styling the components properly, I'll keep you posted.

**Comment by bklebe** [comment]
> your link doesn't have the search term quoted

That was intentional because some are def and some are defp, but with the benefit of hindsight, here's a less stupid search: https://github.com/search?q=repo%3Ambta%2Fdotcom+%28%22def+icon%28%22+OR+%22defp+icon%28%22%29&type=code

**Comment by thecristen** [comment]
@bklebe oh snap! Let's rename it (and any others)

**Review by thecristen — CHANGES_REQUESTED** [review]
I had to add an import of the Tailwind CSS into our CSS to get it to style the components, did you forget it?

1. Import it into the beginning of the `app.scss` file:
```scss
@import "~tailwindcss/base";
@import "~tailwindcss/components";
@import "~tailwindcss/utilities";
```

I also figured out Tailwind apparently includes some opinionated base styles, some of which conflict with ours. So you'll need to do this too:

2. Disable Tailwind's "Preflight" feature with [these instructions](https://tailwindcss.com/docs/preflight#disabling-preflight)

Curious to see this work on a deployment too.

File: lib/dotcom_web/components/core_components.ex [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+    <div
+      :if={msg = render_slot(@inner_block) || Phoenix.Flash.get(@flash, @kind)}
+      id={@id}
+      phx-click={JS.push("lv:clear-flash", value: %{key: @kind}) |> hide("##{@id}")}
+      role="alert"
+      class={[
+        "fixed top-2 right-2 mr-2 w-80 sm:w-96 z-50 rounded-lg p-3 ring-1",
+        @kind == :info && "bg-emerald-50 text-emerald-800 ring-emerald-500 fill-cyan-900",
+        @kind == :error && "bg-rose-50 text-rose-900 shadow-md ring-rose-500 fill-rose-900"
+      ]}
```
Comment (thecristen) [inline lib/dotcom_web/components/core_components.ex:125]: missed some prefixing here!

File: lib/dotcom_web/components/core_components.ex [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+
+  def button(assigns) do
+    ~H"""
+    <button
+      type={@type}
+      class={[
+        "phx-submit-loading:opacity-75 rounded-lg bg-zinc-900 hover:bg-zinc-700 py-2 px-3",
+        "text-sm font-semibold leading-6 text-white active:text-white/80",
+        @class
+      ]}
```
Comment (thecristen) [inline lib/dotcom_web/components/core_components.ex:251]: missed some prefixing here too

File: lib/dotcom_web/components/core_components.ex [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+    <div>
+      <.label for={@id}><%= @label %></.label>
+      <textarea
+        id={@id}
+        name={@name}
+        class={[
+          "mt-2 block w-full rounded-lg text-zinc-900 focus:ring-0 sm:text-sm sm:leading-6 min-h-[6rem]",
+          @errors == [] && "border-zinc-300 focus:border-zinc-400",
+          @errors != [] && "border-rose-400 focus:border-rose-400"
+        ]}
```
Comment (thecristen) [inline lib/dotcom_web/components/core_components.ex:375]: more prefixing!

File: lib/dotcom_web/components/core_components.ex [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+      <input
+        type={@type}
+        name={@name}
+        id={@id}
+        value={Phoenix.HTML.Form.normalize_value(@type, @value)}
+        class={[
+          "mt-2 block w-full rounded-lg text-zinc-900 focus:ring-0 sm:text-sm sm:leading-6",
+          @errors == [] && "border-zinc-300 focus:border-zinc-400",
+          @errors != [] && "border-rose-400 focus:border-rose-400"
+        ]}
```
Comment (thecristen) [inline lib/dotcom_web/components/core_components.ex:397]: missing prefixes

File: lib/dotcom_web/components/core_components.ex [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+  attr :class, :string, default: nil
+
+  slot :inner_block, required: true
+  slot :subtitle
+  slot :actions
+
+  def header(assigns) do
+    ~H"""
+    <header class={[@actions != [] && "flex items-center justify-between gap-6", @class]}>
+      <div>
```
Comment (thecristen) [inline lib/dotcom_web/components/core_components.ex:445]: Missing prefixes here

File: lib/dotcom_web/components/core_components.ex [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+        <tbody
+          id={@id}
+          phx-update={match?(%Phoenix.LiveView.LiveStream{}, @rows) && "stream"}
+          class="tw-relative tw-divide-y tw-divide-zinc-100 tw-border-t tw-border-zinc-200 tw-text-sm tw-leading-6 tw-text-zinc-700"
+        >
+          <tr :for={row <- @rows} id={@row_id && @row_id.(row)} class="tw-group tw-hover:bg-zinc-50">
+            <td
+              :for={{col, i} <- Enum.with_index(@col)}
+              phx-click={@row_click && @row_click.(row)}
+              class={["relative p-0", @row_click && "hover:cursor-pointer"]}
```
Comment (thecristen) [inline lib/dotcom_web/components/core_components.ex:511]: missed prefix here

File: lib/dotcom_web/components/core_components.ex [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+            >
+              <div class="tw-block tw-py-4 tw-pr-6">
+                <span class="tw-absolute tw--inset-y-px tw-right-0 tw--left-4 tw-group-hover:bg-zinc-50 tw-sm:rounded-l-xl" />
+                <span class={["relative", i == 0 && "font-semibold text-zinc-900"]}>
```
Comment (thecristen) [inline lib/dotcom_web/components/core_components.ex:515]: Missed prefix here

File: lib/dotcom_web/components/core_components.ex [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+  ## JS Commands
+
+  def show(js \\ %JS{}, selector) do
+    JS.show(js,
+      to: selector,
+      time: 300,
+      transition:
+        {"transition-all transform ease-out duration-300",
+         "opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95",
+         "opacity-100 translate-y-0 sm:scale-100"}
```
Comment (thecristen) [inline lib/dotcom_web/components/core_components.ex:625]: I think these are all CSS class names? They need prefixes

File: lib/dotcom_web/components/core_components.ex [hunk 9]
```diff
@@ -0,0 +1,692 @@
...
+  def show_modal(js \\ %JS{}, id) when is_binary(id) do
+    js
+    |> JS.show(to: "##{id}")
+    |> JS.show(
+      to: "##{id}-bg",
+      time: 300,
+      transition: {"transition-all transform ease-out duration-300", "opacity-0", "opacity-100"}
+    )
+    |> show("##{id}-container")
+    |> JS.add_class("overflow-hidden", to: "body")
```
Comment (thecristen) [inline lib/dotcom_web/components/core_components.ex:649]: missing prefixes

File: lib/dotcom_web/templates/page/index.html.eex [hunk 10]
```diff
@@ -1,4 +1,4 @@
-<div class="index">
+<div class="index tw-green">
```
Comment (thecristen) [inline lib/dotcom_web/templates/page/index.html.eex:1]: Hahaha I didn't even notice this, it should've been way obvious 😂

Comment (thecristen) [reply lib/dotcom_web/templates/page/index.html.eex:1]: But I suppose `.green` isn't a real Tailwind class anyhow. Let's remove it.

**Review by thecristen — CHANGES_REQUESTED** [review]
Just a few spots in the core components that missed the Tailwind class prefixing!

Works locally but doesn't seem to the working for the prod build, feel free to use dev-blue to test out stuff!

File: lib/dotcom_web/templates/page/index.html.eex (hunk 10 above)
Comment (thecristen) [reply lib/dotcom_web/templates/page/index.html.eex:1]: Oh, didn't realize you just added this, can disregard 😅

**Review by thecristen — APPROVED** [review]
FYI it's hard to tell what's changed between each of my reviews -- I recommend not squash+force-pushing after reviewing's begun. Anyhow, from what I can tell all the pieces are in place! 💯

**Pull request merged**